eventsocket/eventsocket.go

```diff
@@ -45,6 +49,63 @@ type Connection struct {
 	textreader    *textproto.Reader
 	err           chan error
 	cmd, api, evt chan *Event
+	hub           *RequestHub
+}
+
+// Request is a bgapi request struct
+type Request struct {
+	UUID      string
+	timestamp time.Time
+	resp      chan *Event
+	err       chan error
+	command   string
+}
+
+// RequestHub is the asyn request/reply hub for bgapi commands
+type RequestHub struct {
+	requests  map[string]*Request
+	inbound   chan *Request
+	responses chan *Event
+	timeout   chan bool
+}
+
+func (h *RequestHub) run() {
+	go func() {
+		for {
+			time.Sleep(1 * time.Second)
+			h.timeout <- true
+		}
+	}()
+	go func() {
+		for {
+			select {
+			case request := <-h.inbound:
+				h.requests[request.UUID] = request
+			case response := <-h.responses:
+				req, ok := h.requests[response.Header["Job-Uuid"].(string)]
+				if ok != true {
+					break
+				}
+				reply := response.Body
+				if strings.HasPrefix(reply, "-E") {
+					if len(reply) > 5 {
+						req.err <- errors.New(reply[5:])
+					} else {
+						req.err <- errors.New(reply)
+					}
+					break
+				}
+				req.resp <- response
+				delete(h.requests, req.UUID)
+			case _ = <-h.timeout:
+				for _, req := range h.requests {
+					if time.Since(req.timestamp) > (time.Millisecond * requestTimeout) {
```

Comment:
I think the request needs to be deleted from the map after it times out. If it isn't, looks like when the response come back it's going to hang forever trying to send to the request's resp channel. And if it never comes back the it's a memory leak.